File: Android/app/src/main/java/no/artsdatabanken/artsorakel/service/RssFeedService.kt
package no.artsdatabanken.artsorakel.service

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import no.artsdatabanken.artsorakel.BuildConfig
import no.artsdatabanken.artsorakel.model.RssFeedItem
import no.artsdatabanken.artsorakel.model.RssCategory
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.net.URL
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RssFeedService @Inject constructor(
    @param:ApplicationContext private val context: Context
) {
    private val prefs: SharedPreferences by lazy {
        context.getSharedPreferences("rss_prefs", Context.MODE_PRIVATE)
    }

    private companion object {
        const val KEY_DISMISSED_GUIDS = "dismissed_guids"
        const val KEY_FEED_FETCHED_SESSION = "feed_fetched_session"
        const val KEY_SESSION_ID = "session_id"
    }

    private val sessionId: String by lazy {
        val savedSessionId = prefs.getString(KEY_SESSION_ID, null)
        val currentTime = System.currentTimeMillis().toString()
        if (savedSessionId == null) {
            prefs.edit().putString(KEY_SESSION_ID, currentTime).apply()
            currentTime
        } else {
            savedSessionId
        }
    }

    private var cachedAllItems: List<RssFeedItem>? = null

    suspend fun fetchFilteredRssItem(): RssFeedItem? = withContext(Dispatchers.IO) {
        try {
            val fetchedSession = prefs.getString(KEY_FEED_FETCHED_SESSION, null)
            val allItems = if (fetchedSession == sessionId && cachedAllItems != null) {
                cachedAllItems!!
            } else {
                val items = fetchAllRssItems()
                cachedAllItems = items
                prefs.edit().putString(KEY_FEED_FETCHED_SESSION, sessionId).apply()
                cleanupOldDismissedGuids(items.map { it.guid })
                items
            }

            val dismissedGuids = getDismissedGuids()

            val filteredItems = allItems.filter { item ->
                passesAllFilters(item) && !isDismissed(item, dismissedGuids)
            }

            filteredItems.firstOrNull()
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    fun refilterCachedItems(): RssFeedItem? {
        val allItems = cachedAllItems ?: return null

        val dismissedGuids = getDismissedGuids()

        val filteredItems = allItems.filter { item ->
            passesAllFilters(item) && !isDismissed(item, dismissedGuids)
        }

        return filteredItems.firstOrNull()
    }

    fun resetSessionOnAppStart() {
        val currentTime = System.currentTimeMillis().toString()
        prefs.edit().putString(KEY_SESSION_ID, currentTime).apply()
        cachedAllItems = null
    }

    private suspend fun fetchAllRssItems(): List<RssFeedItem> = withContext(Dispatchers.IO) {
        val items = mutableListOf<RssFeedItem>()

        try {
            val url = URL(BuildConfig.RSS_FEED_URL)
            val connection = url.openConnection()
            connection.connect()
            val inputStream = connection.getInputStream()

            val factory = XmlPullParserFactory.newInstance()
            factory.isNamespaceAware = false
            val parser = factory.newPullParser()
            parser.setInput(inputStream, null)

            var eventType = parser.eventType
            var currentGuid: String? = null
            var currentTitle: String? = null
            var currentDescription: String? = null
            var currentLink: String? = null
            var currentPubDate: String? = null
            var currentLanguage: String? = null
            val currentCategories = mutableListOf<String>()
            var currentRssCategory: RssCategory = RssCategory.INFO
            var insideItem = false
            var insideChannel = false
            var channelLanguage: String? = null

            while (eventType != XmlPullParser.END_DOCUMENT) {
                when (eventType) {
                    XmlPullParser.START_TAG -> {
                        when (parser.name?.lowercase()) {
                            "channel" -> insideChannel = true
                            "language" -> {
                                if (insideChannel && !insideItem) {
                                    channelLanguage = parser.nextText()
                                }
                            }
                            "item" -> {
                                insideItem = true
                                currentLanguage = channelLanguage
                                currentCategories.clear()
                                currentRssCategory = RssCategory.INFO
                            }
                            "guid" -> if (insideItem) currentGuid = parser.nextText()
                            "title" -> if (insideItem) currentTitle = parser.nextText()
                            "description" -> if (insideItem) currentDescription = parser.nextText()
                            "link" -> if (insideItem) currentLink = parser.nextText()
                            "pubdate" -> if (insideItem) currentPubDate = parser.nextText()
                            "category" -> if (insideItem) {
                                val categoryText = parser.nextText()
                                when (categoryText.uppercase()) {
                                    "DANGER" -> currentRssCategory = RssCategory.DANGER
                                    "WARNING" -> currentRssCategory = RssCategory.WARNING
                                    "INFO" -> currentRssCategory = RssCategory.INFO
                                    else -> currentCategories.add(categoryText)
                                }
                            }
                        }
                    }
                    XmlPullParser.END_TAG -> {
                        when (parser.name?.lowercase()) {
                            "channel" -> {
                                insideChannel = false
                                channelLanguage = null
                            }
                            "item" -> {
                                if (insideItem && currentGuid != null && currentTitle != null && currentDescription != null) {
                                    items.add(
                                        RssFeedItem(
                                            guid = currentGuid,
                                            title = currentTitle,
                                            description = cleanDescription(currentDescription),
                                            link = currentLink,
                                            pubDate = currentPubDate,
                                            language = currentLanguage,
                                            categories = currentCategories.toList(),
                                            category = currentRssCategory
                                        )
                                    )
                                }
                                insideItem = false
                                currentGuid = null
                                currentTitle = null
                                currentDescription = null
                                currentLink = null
                                currentPubDate = null
                                currentLanguage = null
                            }
                        }
                    }
                }
                eventType = parser.next()
            }

            inputStream.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        items
    }

    private fun passesAllFilters(item: RssFeedItem): Boolean {
        return passesLanguageFilter(item) &&
               passesVersionFilter(item) &&
               passesPlatformFilter(item)
    }

    private fun passesLanguageFilter(item: RssFeedItem): Boolean {
        if (item.language == null) return true

        val currentLocale = Locale.getDefault()
        val currentLanguage = currentLocale.language

        return item.language == currentLanguage ||
               (item.language == "nb" && currentLanguage == "no") ||
               (item.language == "no" && currentLanguage == "nb")
    }

    private fun passesVersionFilter(item: RssFeedItem): Boolean {
        val versionRequirements = item.categories.filter { it.startsWith("Version:") }
        if (versionRequirements.isEmpty()) return true

        val currentVersion = BuildConfig.VERSION_NAME

        return versionRequirements.all { requirement ->
            val versionPart = requirement.substring("Version:".length).trim()
            checkVersionRequirement(currentVersion, versionPart)
        }
    }

    private fun checkVersionRequirement(currentVersion: String, requirement: String): Boolean {
        val operators = listOf("<=", ">=", "<", ">", "=")
        var operator = "="
        var versionToCompare = requirement

        for (op in operators) {
            if (requirement.startsWith(op)) {
                operator = op
                versionToCompare = requirement.substring(op.length).trim()
                break
            }
        }

        if (!versionToCompare.contains(Regex("^\\d+(\\.\\d+)*$"))) {
            versionToCompare = requirement
            operator = "="
        }

        val comparison = compareVersions(currentVersion, versionToCompare)

        return when (operator) {
            "<" -> comparison < 0
            "<=" -> comparison <= 0
            ">" -> comparison > 0
            ">=" -> comparison >= 0
            "=" -> comparison == 0
            else -> comparison == 0
        }
    }

    private fun compareVersions(version1: String, version2: String): Int {
        val parts1 = version1.split(".").map { it.toIntOrNull() ?: 0 }
        val parts2 = version2.split(".").map { it.toIntOrNull() ?: 0 }

        val maxLength = maxOf(parts1.size, parts2.size)

        for (i in 0 until maxLength) {
            val part1 = if (i < parts1.size) parts1[i] else 0
            val part2 = if (i < parts2.size) parts2[i] else 0

            if (part1 < part2) return -1
            if (part1 > part2) return 1
        }

        return 0
    }

    private fun passesPlatformFilter(item: RssFeedItem): Boolean {
        val platformRequirements = item.categories.filter { it.startsWith("Platform:") }
        if (platformRequirements.isEmpty()) return true

        val currentPlatform = "Android"

        return platformRequirements.any { requirement ->
            val platform = requirement.substring("Platform:".length).trim()
            platform.equals(currentPlatform, ignoreCase = true)
        }
    }

    private fun isDismissed(item: RssFeedItem, dismissedGuids: Set<String>): Boolean {
        // Permanent items are never considered dismissed since they can't be dismissed
        if (item.isPermanent) return false

        return dismissedGuids.contains(item.guid)
    }

    fun dismissItem(item: RssFeedItem) {
        // Permanent items cannot be dismissed
        if (item.isPermanent) return

        val dismissedGuids = getDismissedGuids().toMutableSet()
        dismissedGuids.add(item.guid)
        saveDismissedGuids(dismissedGuids)
    }

    private fun getDismissedGuids(): Set<String> {
        val guidsString = prefs.getString(KEY_DISMISSED_GUIDS, "") ?: ""
        return if (guidsString.isEmpty()) {
            emptySet()
        } else {
            guidsString.split(",").toSet()
        }
    }

    private fun saveDismissedGuids(guids: Set<String>) {
        prefs.edit().putString(KEY_DISMISSED_GUIDS, guids.joinToString(",")).apply()
    }

    private fun cleanupOldDismissedGuids(currentGuids: List<String>) {
        val dismissedGuids = getDismissedGuids()
        val currentGuidsSet = currentGuids.toSet()
        val cleanedGuids = dismissedGuids.intersect(currentGuidsSet)

        if (cleanedGuids.size < dismissedGuids.size) {
            saveDismissedGuids(cleanedGuids)
        }
    }

    private fun cleanDescription(description: String): String {
        return description
            .replace(Regex("<[^>]*>"), "")
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")
            .trim()
    }
}

File: iOS/Artsorakel/Artsorakel/Services/RssFeedService.swift
//
//  RssFeedService.swift
//  Artsorakel
//

import Foundation

class RssFeedService: ObservableObject {
    static let shared = RssFeedService()

    @Published private(set) var currentFeedItem: RssFeedItem?

    private let userDefaultsKey = "rss_dismissed_guids"
    private let sessionIdKey = "rss_session_id"
    private let feedFetchedSessionKey = "rss_feed_fetched_session"

    private var cachedAllItems: [RssFeedItem]?
    private var sessionId: String

    private init() {
        if let savedSessionId = UserDefaults.standard.string(forKey: sessionIdKey) {
            self.sessionId = savedSessionId
        } else {
            let newSessionId = String(Date().timeIntervalSince1970)
            UserDefaults.standard.set(newSessionId, forKey: sessionIdKey)
            self.sessionId = newSessionId
        }
    }

    func resetSessionOnAppStart() {
        let newSessionId = String(Date().timeIntervalSince1970)
        UserDefaults.standard.set(newSessionId, forKey: sessionIdKey)
        self.sessionId = newSessionId
        cachedAllItems = nil
    }

    @MainActor
    func loadRssFeed() async {
        let item = await fetchFilteredRssItem()
        self.currentFeedItem = item
    }

    @MainActor
    func dismissCurrentItem() {
        guard let item = currentFeedItem else { return }
        dismissItem(item)
        let nextItem = refilterCachedItems()
        self.currentFeedItem = nextItem
    }

    @MainActor
    func onLanguageChanged() {
        let item = refilterCachedItems()
        self.currentFeedItem = item
    }

    private func fetchFilteredRssItem() async -> RssFeedItem? {
        let fetchedSession = UserDefaults.standard.string(forKey: feedFetchedSessionKey)

        let allItems: [RssFeedItem]
        if fetchedSession == sessionId, let cached = cachedAllItems {
            allItems = cached
        } else {
            allItems = await fetchAllRssItems()
            cachedAllItems = allItems
            UserDefaults.standard.set(sessionId, forKey: feedFetchedSessionKey)
            cleanupOldDismissedGuids(currentGuids: allItems.map { $0.id })
        }

        let dismissedGuids = getDismissedGuids()

        let filteredItems = allItems.filter { item in
            passesAllFilters(item) && !isDismissed(item, dismissedGuids: dismissedGuids)
        }

        return filteredItems.first
    }

    private func refilterCachedItems() -> RssFeedItem? {
        guard let allItems = cachedAllItems else { return nil }

        let dismissedGuids = getDismissedGuids()

        let filteredItems = allItems.filter { item in
            passesAllFilters(item) && !isDismissed(item, dismissedGuids: dismissedGuids)
        }

        return filteredItems.first
    }

    private func fetchAllRssItems() async -> [RssFeedItem] {
        guard let urlString = AppConfig.shared.rssFeedURL,
              let url = URL(string: urlString) else {
            return []
        }

        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            return parseRssFeed(data: data)
        } catch {
            return []
        }
    }

    private func parseRssFeed(data: Data) -> [RssFeedItem] {
        let parser = RssXmlParser()
        return parser.parse(data: data)
    }

    // MARK: - Filtering

    private func passesAllFilters(_ item: RssFeedItem) -> Bool {
        return passesLanguageFilter(item) &&
               passesVersionFilter(item) &&
               passesPlatformFilter(item)
    }

    private func passesLanguageFilter(_ item: RssFeedItem) -> Bool {
        guard let itemLanguage = item.language else { return true }

        // Use app's language setting, falling back to device locale
        let savedLanguage = UserDefaults.standard.string(forKey: "selectedLanguage") ?? "system"
        let currentLanguage: String
        if savedLanguage == "system" {
            currentLanguage = Locale.current.languageCode ?? "en"
        } else {
            currentLanguage = savedLanguage
        }

        return itemLanguage == currentLanguage ||
               (itemLanguage == "nb" && currentLanguage == "no") ||
               (itemLanguage == "no" && currentLanguage == "nb")
    }

    private func passesVersionFilter(_ item: RssFeedItem) -> Bool {
        let versionRequirements = item.categories.filter { $0.hasPrefix("Version:") }
        if versionRequirements.isEmpty { return true }

        let currentVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"

        return versionRequirements.allSatisfy { requirement in
            let versionPart = String(requirement.dropFirst("Version:".count)).trimmingCharacters(in: .whitespaces)
            return checkVersionRequirement(currentVersion: currentVersion, requirement: versionPart)
        }
    }

    private func checkVersionRequirement(currentVersion: String, requirement: String) -> Bool {
        let operators = ["<=", ">=", "<", ">", "="]
        var op = "="
        var versionToCompare = requirement

        for operatorStr in operators {
            if requirement.hasPrefix(operatorStr) {
                op = operatorStr
                versionToCompare = String(requirement.dropFirst(operatorStr.count)).trimmingCharacters(in: .whitespaces)
                break
            }
        }

        // Validate version format
        let versionRegex = try? NSRegularExpression(pattern: "^\\d+(\\.\\d+)*$")
        let range = NSRange(versionToCompare.startIndex..., in: versionToCompare)
        if versionRegex?.firstMatch(in: versionToCompare, range: range) == nil {
            versionToCompare = requirement
            op = "="
        }

        let comparison = compareVersions(currentVersion, versionToCompare)

        switch op {
        case "<": return comparison < 0
        case "<=": return comparison <= 0
        case ">": return comparison > 0
        case ">=": return comparison >= 0
        case "=": return comparison == 0
        default: return comparison == 0
        }
    }

    private func compareVersions(_ version1: String, _ version2: String) -> Int {
        let parts1 = version1.split(separator: ".").map { Int($0) ?? 0 }
        let parts2 = version2.split(separator: ".").map { Int($0) ?? 0 }

        let maxLength = max(parts1.count, parts2.count)

        for i in 0..<maxLength {
            let part1 = i < parts1.count ? parts1[i] : 0
            let part2 = i < parts2.count ? parts2[i] : 0

            if part1 < part2 { return -1 }
            if part1 > part2 { return 1 }
        }

        return 0
    }

    private func passesPlatformFilter(_ item: RssFeedItem) -> Bool {
        let platformRequirements = item.categories.filter { $0.hasPrefix("Platform:") }
        if platformRequirements.isEmpty { return true }

        let currentPlatform = "iOS"

        return platformRequirements.contains { requirement in
            let platform = String(requirement.dropFirst("Platform:".count)).trimmingCharacters(in: .whitespaces)
            return platform.caseInsensitiveCompare(currentPlatform) == .orderedSame
        }
    }

    // MARK: - Dismissal

    private func isDismissed(_ item: RssFeedItem, dismissedGuids: Set<String>) -> Bool {
        // Permanent items are never considered dismissed since they can't be dismissed
        if item.isPermanent { return false }
        return dismissedGuids.contains(item.id)
    }

    private func dismissItem(_ item: RssFeedItem) {
        // Permanent items cannot be dismissed
        if item.isPermanent { return }

        var dismissedGuids = getDismissedGuids()
        dismissedGuids.insert(item.id)
        saveDismissedGuids(dismissedGuids)
    }

    private func getDismissedGuids() -> Set<String> {
        guard let guidsString = UserDefaults.standard.string(forKey: userDefaultsKey),
              !guidsString.isEmpty else {
            return []
        }
        return Set(guidsString.split(separator: ",").map { String($0) })
    }

    private func saveDismissedGuids(_ guids: Set<String>) {
        UserDefaults.standard.set(guids.joined(separator: ","), forKey: userDefaultsKey)
    }

    private func cleanupOldDismissedGuids(currentGuids: [String]) {
        let dismissedGuids = getDismissedGuids()
        let currentGuidsSet = Set(currentGuids)
        let cleanedGuids = dismissedGuids.intersection(currentGuidsSet)

        if cleanedGuids.count < dismissedGuids.count {
            saveDismissedGuids(cleanedGuids)
        }
    }
}

// MARK: - XML Parser

private class RssXmlParser: NSObject, XMLParserDelegate {
    private var items: [RssFeedItem] = []

    private var currentGuid: String?
    private var currentTitle: String?
    private var currentDescription: String?
    private var currentLink: String?
    private var currentPubDate: String?
    private var currentLanguage: String?
    private var currentCategories: [String] = []
    private var currentRssCategory: RssCategory = .info

    private var insideItem = false
    private var insideChannel = false
    private var channelLanguage: String?
    private var currentElement: String = ""
    private var currentText: String = ""

    func parse(data: Data) -> [RssFeedItem] {
        items = []
        let parser = XMLParser(data: data)
        parser.delegate = self
        parser.parse()
        return items
    }

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String: String] = [:]) {
        currentElement = elementName.lowercased()
        currentText = ""

        switch currentElement {
        case "channel":
            insideChannel = true
        case "item":
            insideItem = true
            currentLanguage = channelLanguage
            currentCategories = []
            currentRssCategory = .info
            currentGuid = nil
            currentTitle = nil
            currentDescription = nil
            currentLink = nil
            currentPubDate = nil
        default:
            break
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        currentText += string
    }

    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        let element = elementName.lowercased()
        let text = currentText.trimmingCharacters(in: .whitespacesAndNewlines)

        switch element {
        case "channel":
            insideChannel = false
            channelLanguage = nil
        case "language":
            if insideChannel && !insideItem {
                channelLanguage = text
            }
        case "item":
            if insideItem,
               let guid = currentGuid,
               let title = currentTitle,
               let description = currentDescription {
                let item = RssFeedItem(
                    guid: guid,
                    title: title,
                    description: cleanDescription(description),
                    link: currentLink,
                    pubDate: currentPubDate,
                    language: currentLanguage,
                    categories: currentCategories,
                    category: currentRssCategory
                )
                items.append(item)
            }
            insideItem = false
        case "guid":
            if insideItem { currentGuid = text }
        case "title":
            if insideItem { currentTitle = text }
        case "description":
            if insideItem { currentDescription = text }
        case "link":
            if insideItem { currentLink = text }
        case "pubdate":
            if insideItem { currentPubDate = text }
        case "category":
            if insideItem {
                let upperText = text.uppercased()
                switch upperText {
                case "DANGER":
                    currentRssCategory = .danger
                case "WARNING":
                    currentRssCategory = .warning
                case "INFO":
                    currentRssCategory = .info
                default:
                    currentCategories.append(text)
                }
            }
        default:
            break
        }

        currentElement = ""
        currentText = ""
    }

    private func cleanDescription(_ description: String) -> String {
        var result = description
        // Remove HTML tags
        if let regex = try? NSRegularExpression(pattern: "<[^>]*>", options: []) {
            result = regex.stringByReplacingMatches(in: result, options: [], range: NSRange(result.startIndex..., in: result), withTemplate: "")
        }
        // Decode HTML entities
        result = result.replacingOccurrences(of: "&nbsp;", with: " ")
        result = result.replacingOccurrences(of: "&amp;", with: "&")
        result = result.replacingOccurrences(of: "&lt;", with: "<")
        result = result.replacingOccurrences(of: "&gt;", with: ">")
        result = result.replacingOccurrences(of: "&quot;", with: "\"")
        result = result.replacingOccurrences(of: "&#39;", with: "'")
        return result.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
